좋아요 관련 기능 구현

dldmldlsy opened the pull request:
개발 목록
--
1. 좋아요 등록 기능 구현
2. 좋아요 조회 기능 구현
3. 좋아요 삭제 기능 구현
4. Controller, Service, Repository Unit 테스트 코드 작성 완료
5. LikeRestControllerDocsTest 작성 완료
6. 좋아요 관련 API 문서 작성 완료

**[추후 개발 목록]** 
- [x] 좋아요 카운트 관련 로직 추가
- [x] Controller 테스트 코드 작성
- [x] RestDocs 코드 작성

dldmldlsy committed:
feat: like request, response dto 구현

dldmldlsy committed:
feat: like create api 기능 구현
LiekRestcontroller-createLike(): 회원정보와 여행정보 받아서 좋아요 등록하는 기능
LikeService-createLike(): 회원정보와 여행정보 받아서 좋아요 생성 후 저장
LikeNotFoundException&LikeRestControllerAdvice-likeNotFoundException(): 좋아요정보를 찾을 수 없을 경우의 예외처리
LikeEntity에 @Builder를 이용한 생성자 추가

dldmldlsy committed:
feat: like get api 기능 구현
LikeRestController-getLikeByMemberIdAndTripId(): 회원id와 여행id 받아서 좋아요 정보 응답하는 controller
LikeService-getLikeByMemberIdAndTripId(): 회원id와 여행id를 이용해 좋아요 entity 조회 후 좋아요 정보 응답 dto 반환하는 기능 구현(해당하는 좋아요정보 없을 시 LikeNotFoundException 발생)
LikeRepository-findByMemberIdAndTripId(): 회원id와 여행id를 이용하여 Like Entity 조회하는 메서드

dldmldlsy committed:
feat: like delete api 기능 구현
LikeRestController-deleteLikeById(): like id를 받아서 삭제 후 삭제한 좋아요 정보 보여주는 controller
LikeService-deleteLikeById(): like id로 Like Entity 조회하여 해당 entity 삭제 후 삭제한 좋아요 정보 응답 dto 반환하는 기능 구현(해당 좋아요 정보가 없을 경우 LikeNotFoundException 발생)

YurimYang reviewed (commented):
코드 확인했습니다 ! 테스트 코드는 아직 안올리신걸까요 ?

dldmldlsy committed:
feat: like 관련 테스트 코드 작성
좋아요 등록 성공
좋아요 조회 성공
좋아요 조회 실패: 해당 id를 가진 좋아요 정보 없음
좋아요 삭제 성공
좋아요 삭제 실패: 해당 id를 가진 좋아요 정보 없음

dldmldlsy commented:
> 코드 확인했습니다 ! 테스트 코드는 아직 안올리신걸까요 ?

네네 테스트코드에 문제가 있어서ㅠ.ㅠ 이전에는 문제없는지 확인하고자 메인코드 먼저 올렸습니다!

dldmldlsy committed:
fix: 테스트코드 수정
좋아요 삭제 시 delete ->  deleteById로 변경
잘못된 메서드명 및 로직 수정

dldmldlsy committed:
fix: 좋아요 관련 service 테스트 코드 수정
문제: createLike 과정에서 쓰이는 tripService에 대해서 Mock객체 생성을 안함

dldmldlsy committed:
feat: LikeRepository 테스트코드 작성
findByMemberIdAndTripId()에 대한 테스트코드 작성

dldmldlsy committed:
fix: LikeRepository 테스트코드 수정
@InjectMocks -> @Mock으로 변경

dldmldlsy committed:
feat: Like 응답 dto 수정
entity->dto 변환을 위한 생성자에 불필요한 어노테이션(@Builder) 삭제
@Builder를 이용한 LikeResponseDTO 생성자 추가 구현

dldmldlsy committed:
feat: 좋아요 등록, 조회, 삭제 API문서화를 위한 테스트 코드 작성

dldmldlsy committed:
fix: 좋아요API 문서화를 위한 테스트코드 수정
조회: requestBody-> pathVariable/ 메서드 명 수정
삭제: uri 수정 /메서드명 수정
등록: Mock객체에 값 넣을때 파라미터 타입 통일 (any->진짜값)

dldmldlsy committed:
docs: 좋아요 등록, 조회, 삭제 API DOCS 작성
fix: LikeRestControllerDocsTest 내 메서드명 수정

dldmldlsy committed:
feat: 좋아요 삭제 시, 좋아요를 등록한 회원이 맞는지 검증 로직 추가
LikeUnauthorizedException 발생

dldmldlsy committed:
feat: 좋아요 repository 테스트코드 고도화
Mock객체 이용한 테스트가 아닌, 실제 H2 DB에 접근하여 테스트 진행
DataJpaTest 활용 (Config설정)
save, findById, findByMemberIdAndTripId, deleteById에 대해 테스트 수행

dldmldlsy committed:
Merge branch 'develop' into feature/like-create-get-delete-api

dldmldlsy committed:
feat: 좋아요 등록/삭제 시 회원정보에 좋아요 수  증가/감소 로직 추가

JeongUijeong reviewed (commented):
수고하셨어요! 테스트 코드 수정하시고 모두 통과되면 approve 해드릴게용

YurimYang commented on src/main/java/com/fc/toy_project3/domain/like/service/LikeService.java:
```diff
@@ -0,0 +1,84 @@
+     * @param likeId 조회할 좋아요 ID
+     * @return 좋아요 정보 Entity
+     */
+    public Like getLike(Long likeId){
```
이부분은 deleteLikeById에 넣으셔도 좋을 것 같습니다 :)

dldmldlsy replied:
다른 서비스처럼 getEntity를 만든 건데 지금은 delete에서밖에 안쓰여서 그런걸까용??

YurimYang replied:
Like like = getLike(likeId);
Like like = likeRepository.findById(likeId).orElseThrow(LikeNotFoundException::new);
이렇게 적어도 되지 않나해서요! getLike가 따로 빠진 이유가 있을까요 ?

dldmldlsy replied:
넵! 그렇게 적어도 됩니다! 지금은 delete에서밖에 안쓰이지만 만약에 likeId로 조회하는 함수가 필요하거나 다른 로직이 생길때 활용될 수 있을 것 같아서 빼서 만든 거였습니당 ㅎㅎ  그럼 그렇게 수정해서 올릴게용

dldmldlsy committed:
feat: 좋아요 controller 테스트코드 작성
성공케이스만 작성

dldmldlsy committed:
fix: 좋아요 테스트코드 에러 수정
좋아요 요청 dto에 생성자 어노테이션 추가
repository 테스트코드에서 SQL 오류 수정 (like->likes)

dldmldlsy committed:
fix: 좋아요 관련 테스트코드 수정
좋아요 응답 dto 생성자 어노테이션 추가
likes 테이블의 필드명 오류 likeId -> id
회원 혹은 여행 저장 시, 이미 있는 정보인지 확인 후에 저장

dldmldlsy committed:
fix: 좋아요 관련 테스트코드 수정

dldmldlsy committed:
fix: 좋아요 관련 테스트코드 수정

dldmldlsy committed:
refactor: 불필요한 메서드 삭제
getLike 함수내용 deleteLikeById에 바로 반영 후 getLike 함수 삭제

YurimYang reviewed (approved):
수고하셨습니다 ~!!

JeongUijeong commented on src/main/java/com/fc/toy_project3/domain/like/dto/request/LikeRequestDTO.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@Builder
```
빌더 생성자 패턴으로 통일 부탁드릴게요

dldmldlsy replied:
앗 넵!!!

JeongUijeong commented on src/main/java/com/fc/toy_project3/domain/like/repository/LikeRepository.java:
```diff
@@ -4,5 +4,5 @@
 public interface LikeRepository extends JpaRepository<Like, Long> {
-
-}
+    Like findByMemberIdAndTripId(Long memberId, Long tripId);
```
Optional 로 받지 않으면 찾지 못했을 때 에러 처리 되지 않나요? 괜찮으신가요?

dldmldlsy replied:
그래서 service에서 null값일 때 예외처리되도록 서비스에서 구현했습니다! 
Optional로 바꿔서 orElseThrow로 처리할까용??

JeongUijeong replied:
그게 좋을 것 같아요!

JeongUijeong commented on src/test/java/com/fc/toy_project3/config/TestJpaConfig.java:
해당 파일 변경 사항이 없으니 PR에 올라오지 않도록 해주세요!

JeongUijeong reviewed (changes requested):
수고하셨습니다! 테스트 코드에 실패하는 경우에 대한 코드가 부족한 것 같습니다! 추후 추가해주시면 좋을 것 같고, 이 외에도 몇 가지 코멘트 달아드렸으니 확인부탁드려요!

dldmldlsy committed:
feat: 좋아요 요청 dto 생성자구현 및 조회쿼리 Opional 활용

dldmldlsy committed:
fix:Optional로 인한 에러 수정
